deployment/address_book.go

```diff
@@ -139,16 +162,61 @@ func (m *AddressBookMap) Merge(ab AddressBook) error {
 	if err != nil {
 		return err
 	}
-	for chain, chainAddresses := range addresses {
-		for address, typeAndVersions := range chainAddresses {
-			if err := m.Save(chain, address, typeAndVersions); err != nil {
-				return err
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	for chainSelector, chainAddresses := range addresses {
+		for address, typeAndVersion := range chainAddresses {
+			if _, exists := m.AddressesByChain[chainSelector]; !exists {
+				m.AddressesByChain[chainSelector] = make(map[string]TypeAndVersion)
+			}
+			if _, exists := m.AddressesByChain[chainSelector][address]; exists {
+				return fmt.Errorf("address %s already exists for chain %d", address, chainSelector)
+			}
+			m.AddressesByChain[chainSelector][address] = typeAndVersion
+		}
+	}
+	return nil
+}
+
+// Remove removes the address book addresses specified via the argument from the AddressBookMap.
+// Errors if all the addresses in the given address book are not contained in the AddressBookMap.
+func (m *AddressBookMap) Remove(ab AddressBook) error {
+	addresses, err := ab.Addresses()
+	if err != nil {
+		return err
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	addressNotFound := true
+	for chainSelector, chainAddresses := range addresses {
+		for address, _ := range chainAddresses {
+			if _, exists := m.AddressesByChain[chainSelector][address]; exists {
+				addressNotFound = false
+				delete(m.AddressesByChain[chainSelector], address)
 			}
 		}
 	}
+
+	if addressNotFound {
```

Comment:
if I specify an address book to be removed where one address is present and one is not, I think this will error but still remove the present address. I think a user would expect all or nothing, i.e. either the whole book is removed and no error, or non of the book was removed. To do that you probably need to allocate a new map and only assign it to `m.AddressesByChain` if its successful.

Reply:
Sorry let me clarify. I would expect this to error because the _whole_ address book doesn't exist in the existing one:
```
func TestAddressBook_Remove2(t *testing.T) {
	onRamp100 := NewTypeAndVersion("OnRamp", Version1_0_0)
	addr1 := common.HexToAddress("0x1").String()
	addr2 := common.HexToAddress("0x2").String()
	addr3 := common.HexToAddress("0x3").String()

	ab1 := NewMemoryAddressBookFromMap(map[uint64]map[string]TypeAndVersion{
		chainsel.TEST_90000001.Selector: {
			addr1: onRamp100,
			addr2: onRamp100,
		},
	})
	ab2 := NewMemoryAddressBookFromMap(map[uint64]map[string]TypeAndVersion{
		chainsel.TEST_90000001.Selector: {
			addr2: onRamp100,
			addr3: onRamp100,
		},
	})
	// This should error because the client 
	// asked to remove addr3 which doesn't exist in ab1
	// How would they know that it wasn't removed?
	require.Error(t, ab1.Remove(ab2))
}
```

Reply:
fixed